communication/src/AsyncTlsConnection.cpp

```diff
@@ -384,103 +385,60 @@ bool AsyncTlsConnection::verifyCertificateServer(asio::ssl::verify_context& ctx)
   if (X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT != X509_STORE_CTX_get_error(ctx.native_handle())) {
     return false;
   }
-  static constexpr size_t SIZE = 512;
-  std::string subject(SIZE, 0);
   X509* cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
   if (!cert) {
     LOG_WARN(logger_, "No certificate from client");
     return false;
   }
-  X509_NAME_oneline(X509_get_subject_name(cert), subject.data(), SIZE);
-  auto [valid, peer_id] = checkCertificate(cert, "client", std::string(subject), std::nullopt);
+  auto [valid, peer_id] = checkCertificate(cert, std::nullopt);
   peer_id_ = peer_id;
   return valid;
 }
 
 std::pair<bool, NodeNum> AsyncTlsConnection::checkCertificate(X509* receivedCert,
-                                                              std::string connectionType,
-                                                              const std::string& subject,
                                                               std::optional<NodeNum> expectedPeerId) {
-  // First, perform a basic sanity test, in order to eliminate a disk read if the certificate is
-  // unknown.
-  //
-  // The certificate must have a node id, as we put it in `OU` field on creation.
-  //
-  // Since we use pinning we must know who the remote peer is.
-  // `peerIdPrefixLength` stands for the length of 'OU=' substring
-  int peerIdPrefixLength = 3;
-  std::regex r("OU=\\d*", std::regex_constants::icase);
-  std::smatch sm;
-  regex_search(subject, sm, r);
-  if (sm.length() <= peerIdPrefixLength) {
-    LOG_WARN(logger_, "OU not found or empty: " << subject);
-    return std::make_pair(false, 0);
-  }
-
-  auto remPeer = sm.str().substr(peerIdPrefixLength, sm.str().length() - peerIdPrefixLength);
-  if (0 == remPeer.length()) {
-    LOG_WARN(logger_, "OU empty " << subject);
-    return std::make_pair(false, 0);
-  }
-
-  NodeNum remotePeerId;
-  try {
-    remotePeerId = stoul(remPeer, nullptr);
-  } catch (const std::invalid_argument& ia) {
-    LOG_WARN(logger_, "cannot convert OU, " << subject << ", " << ia.what());
-    return std::make_pair(false, 0);
-  } catch (const std::out_of_range& e) {
-    LOG_WARN(logger_, "cannot convert OU, " << subject << ", " << e.what());
-    return std::make_pair(false, 0);
-  }
-
-  // If the server has been verified, check that the peers match.
-  if (expectedPeerId) {
-    if (remotePeerId != expectedPeerId) {
-      LOG_WARN(logger_, "Peers don't match, expected: " << expectedPeerId.value() << ", received: " << remPeer);
-      return std::make_pair(false, remotePeerId);
+  uint32_t peerId = UINT32_MAX;
+  std::string conn_type;
+  bool res = concord::util::crypto::CertificateUtils::verifyCertificate(
+      receivedCert, config_.certificatesRootPath, peerId, conn_type);
+  if (expectedPeerId.has_value() && peerId != expectedPeerId.value()) return std::make_pair(false, peerId);
+  if (!res) {
```

Comment:
Suggestion: We can return right here and therefore skip the indentation.
```
if (res)  return std::make_pair(res, peer_id);
```

Reply:
fixed